### DISCUSSION TITLE
playing with pyvista celestial bodies

### DISCUSSION THREAD

**Discussion by ifermion** [pyvista/pyvista#4086] [Q&A]
I've been playing with the examples available for visualising 3D celestial bodies in pyvista at the following link.

https://docs.pyvista.org/examples/99-advanced/planets.html

I wanted to visualise some vectors of interest for Earth, and I plotted them in pyvista. It seems to me that the map of the Earth overlain on the 3D mesh is not properly oriented when I plotted x,y and z axes of an Earth-centric cartesian coordinate system. I expected x-axis to go through the intercept of equator and the meridian line (0° longitude) going through the city Greenwich in UK. However it seems to me that there is an angular shift of about 180° between the vector of the x-axis and where it ought to be on the map. Has someone come across this before or am I introducing a glitch somewhere ?

I've giving below the code, I used and a resulting visual.
Thanks in advance.

```python
import numpy as np
import pyvista as pv
from pyvista import examples

plotter = pv.Plotter(window_size=[2100,1400])
plotter.set_background('black')

Earth = examples.planets.load_earth(radius=6378.1)


body_x = np.array([1,0,0])
body_y = np.array([0,1,0])
body_z = np.array([0,0,1])

e_p = np.array([0,0,0])

body_fixed_X_vec = pv.Arrow(start=e_p, direction=body_x*1e4,scale='auto')
body_fixed_Y_vec = pv.Arrow(start=e_p, direction=body_y*1e4,scale='auto')
body_fixed_Z_vec = pv.Arrow(start=e_p, direction=body_z*1e4,scale='auto')

plotter.add_mesh(Earth)

plotter.add_mesh(body_fixed_X_vec,color='blue',opacity=0.7)
plotter.add_mesh(body_fixed_Y_vec,color='yellow',opacity=0.7)
plotter.add_mesh(body_fixed_Z_vec,color='green',opacity=0.7)

plotter.show()

```
Here, blue is the x-axis and yellow is the y-axis of an Earth-centred / geocentric cartesian coordinate system.
<img width="517" alt="image" src="https://user-images.githubusercontent.com/24443359/222921439-73a75422-7ca2-4af4-91aa-c79f1e6bc52d.png">

**Comment by tkoyama010** [comment]
Thanks for the discussion.
We were not concerned about Greenwich. 
We better fix it.

**Reply by ifermion** [reply]
Thanks for your reply. So has this been fixed now?

**Reply by akaszynski** [reply]
Hi @ifermion,

If you can give me a Z rotation in degrees, I can rotate it by default and then add your example to our docs.

**Reply by ifermion** [reply]
so, if I understand correctly, the overlain map should be rotated 180° anticlockwise around the +Z axis. Once the rotation is finished, the blue vector should pop out of the globe from the Greenwich city. 

However, I invite anyone else also to verify this calculation.

Thanks !

**Reply by dcd2** [reply]
My testing confirms @ifermion issue that the texture is not aligned properly. 

@akaszynski - How would one go about rotating the texture? Teach a man to fish and all. I've played around with the code from [Issue 257](https://github.com/pyvista/pyvista-support/issues/257), but to no avail.

The texture should be rotated so that Greenwich is aligned along the Prime Meridian. It would make the textured earth model far more useful for any application using the earth model in ECEF coordinate system described here: https://en.wikipedia.org/wiki/Earth-centered,_Earth-fixed_coordinate_system

As it stands, marking positions based on ECEF coordinates results in the coordinates being placed on the "wrong" spot on the globe (its the texture that's off, not the position...I think).

**Reply by dcd2** [reply]
I modified @ifermion code a little to show Greenwich's location on the globe using ECEF coordinates (red sphere) vs. it's antipode (orange sphere):


![image](https://github.com/user-attachments/assets/044af71e-bf85-4428-8e0c-80156bbf350f)



In latitude and longitude, Greenwich lies at 51.4934N, 0.0098E, whereas it's antipode is at 51.4934N, 179.9902E. So, the amount of rotation is about 180 degrees as @ifermion mentioned.


Here's the code:

https://github.com/pyvista/pyvista/discussions/4086
https://github.com/pyvista/pyvista-support/issues/257

``` python
import numpy as np
import pyvista as pv
from pyvista import examples

plotter = pv.Plotter(window_size=[2100,1400])
plotter.set_background('black')

earth_semi_major_radius = 6378137

Earth = examples.planets.load_earth(radius=earth_semi_major_radius) # meters
earth_texture = examples.load_globe_texture()


body_x = np.array([1,0,0])
body_y = np.array([0,1,0])
body_z = np.array([0,0,1])

e_p = np.array([0,0,0])

body_fixed_X_vec = pv.Arrow(start=e_p, direction=body_x * earth_semi_major_radius * 2, scale='auto')
body_fixed_Y_vec = pv.Arrow(start=e_p, direction=body_y * earth_semi_major_radius * 2, scale='auto')
body_fixed_Z_vec = pv.Arrow(start=e_p, direction=body_z * earth_semi_major_radius * 2, scale='auto')

#ECEF coordinates
point_merdian = np.array([3.97922312e+06, 6.80615348e+02, 4.96790531e+06])
point_merdian_antipode = np.array([-3.97922312e+06,  6.80615348e+02,  4.96790531e+06])

plotter.add_mesh(Earth, texture=earth_texture)

plotter.add_axes(line_width=5, labels_off=False, color='white')

plotter.add_mesh(body_fixed_X_vec,color='red',opacity=0.7)
plotter.add_mesh(body_fixed_Y_vec,color='green',opacity=0.7)
plotter.add_mesh(body_fixed_Z_vec,color='blue',opacity=0.7)

plotter.add_points(
    point_merdian, render_points_as_spheres=True, point_size=40.0, color='red', opacity=0.5)
plotter.add_points(
    point_merdian_antipode, render_points_as_spheres=True, point_size=40.0, color='orange', opacity=0.5)

plotter.show()
```

**Reply by user27182** [reply]
It looks like the correct orientation is when the UK is approximately aligned with the +X direction.

If you take a look at the [`load_globe`](https://dev.pyvista.org/api/examples/_autosummary/pyvista.examples.examples.load_globe.html#pyvista.examples.examples.load_globe) example instead of [`load_earth`](https://dev.pyvista.org/api/examples/_autosummary/pyvista.examples.planets.load_earth.html#pyvista.examples.planets.load_earth), `load_globe` seems to have the correct orientation.

The difference between these two is `load_globe` loads a dataset from file, where the sphere has quads and defines texture coordinates. By contrast `load_earth` uses `pv.Sphere()` to make the sphere (which has triangle cells) and also does not define texture coordinates.

So a simple workaround may be to simply use `load_globe` instead.

A more proper solution may be to have a `transform_texture_coordinates` method for meshes that can shift or rotate texture coordinates.

**Reply by user27182** [reply]
Here's a working example using `planets.load_earth` with a texture transformation method.

``` python
from pyvista import examples
import numpy as np

def transform_texture_coordinates(mesh, translation=(0, 0), rotation=0, scale=(1, 1), origin=(0.5, 0.5)):
    """Apply a transformation to the texture coordinates of a mesh.

    Parameters
    ----------
    mesh : PolyData
        Mesh with texture coordinates to transform.
    translation : tuple[float, float]
        (tx, ty) translation of texture coordinates.
    rotation : float
        Rotation of texture coordinates in degrees.
    scale : tuple[float, float]
        (sx, sy) scaling of texture coordinates.
    origin : tuple[float, float]
        (ox, oy) rotation origin in texture space. Default is texture center.

    Returns
    -------
    PolyData
        Mesh with transformed texture coordinates.
    """
    # Get texture coordinates
    tcoords = mesh.active_texture_coordinates
    if tcoords is None:
        raise ValueError("Mesh has no texture coordinates.")

    # Translate to origin
    tcoords -= origin

    # Apply scaling
    tcoords *= scale

    # Apply rotation
    if rotation != 0:
        rad = np.radians(rotation)
        cos_theta, sin_theta = np.cos(rad), np.sin(rad)
        rotation_matrix = np.array([[cos_theta, -sin_theta], [sin_theta, cos_theta]])
        tcoords = tcoords @ rotation_matrix.T

    # Translate back from origin and apply final translation
    tcoords += origin
    tcoords += translation

    # Update the mesh with new texture coordinates
    mesh.active_texture_coordinates = tcoords

    return mesh

dataset = examples.planets.load_earth()
texture = examples.load_globe_texture()

# Shift x-coordinates of texture by half to "rotate" it 180 degrees
transformed = transform_texture_coordinates(dataset, translation=(0.5, 0), rotation=0, scale=(1, 1))
transformed.plot(texture=texture)
```

<img src="https://github.com/user-attachments/assets/6136250d-f9cf-42d7-bdee-c92fe7763e36" width="500" />

**Reply by dcd2** [reply]
Thanks for looking into this @user27182.

Yes, the UK, and specifically Greenwich, should line up along the X-axis. 

I'll take a look at `load_globe` as suggested.  I've probably done something wrong, or have an outdated installation, but I received the following error running your working example above:

> runfile('/home/lynx/Downloads/untitled13.py', wdir='/home/lynx/Downloads')
> Traceback (most recent call last):
> 
>   File ~/miniconda3/envs/intersect/lib/python3.11/site-packages/spyder_kernels/py3compat.py:356 in compat_exec
>     exec(code, globals, locals)
> 
>   File ~/Downloads/untitled13.py:64
>     transformed = transform_texture_coordinates(dataset, translation=(0.5, 0), rotation=0, scale=(1, 1))
> 
>   File ~/Downloads/untitled13.py:34 in transform_texture_coordinates
>     tcoords = mesh.active_texture_coordinates
> 
>   File ~/miniconda3/envs/intersect/lib/python3.11/site-packages/pyvista/core/dataset.py:132 in __getattr__
>     return super().__getattribute__(item)
> 
> AttributeError: 'PolyData' object has no attribute 'active_texture_coordinates'

Here's my system:

`print(pv.Report())

--------------------------------------------------------------------------------
  Date: Sun Mar 09 21:32:28 2025 EDT

                OS : Linux
            CPU(s) : 8
           Machine : x86_64
      Architecture : 64bit
               RAM : 31.0 GiB
       Environment : IPython
       File system : ext4
        GPU Vendor : Intel
      GPU Renderer : Mesa Intel(R) UHD Graphics 630 (CFL GT2)
       GPU Version : 4.6 (Core Profile) Mesa 24.0.3-1pop1~1711635559~22.04~7a9f319
  MathText Support : False

  Python 3.11.3 | packaged by conda-forge | (main, Apr  6 2023, 08:57:19) [GCC
  11.3.0]

           pyvista : 0.39.1
               vtk : 9.2.6
             numpy : 1.24.3
        matplotlib : 3.7.1
            scooby : 0.7.2
             pooch : v1.7.0
             PyQt5 : 5.15.7
           IPython : 8.13.2
             scipy : 1.10.1
      nest_asyncio : 1.5.6
--------------------------------------------------------------------------------`

**Reply by user27182** [reply]
I see you're using an older version of pyvista. Try `active_t_coords` instead.

https://docs.pyvista.org/api/core/_autosummary/pyvista.datasetattributes.active_t_coords#pyvista.DataSetAttributes.active_t_coords

**Reply by dcd2** [reply]
I can confirm that the supplied texture transform method works. I've tested it with various satellite images and the ~180 rotation appears to do the trick. 

A small note for novice users like myself who come across this in the future: I slightly modified the code to use a copy of the mesh (e.g. `mesh_copy = mesh.copy()`, or else it updates the texture coordinates of the supplied mesh in place. Using a copy is useful if one wants to compare the mesh before and after the texture coordinates are transformed.

Many thanks!

**Comment by Coleslaw25** [comment]
Found the same issue: https://github.com/pyvista/pyvista/issues/8703#issuecomment-4512931956
PR is up with a fix here: https://github.com/pyvista/pyvista/pull/8705
